orttraining/orttraining/core/framework/ortmodule_graph_builder.cc

```diff
@@ -78,14 +79,28 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
     SetConcreteInputShapes(*input_shapes_ptr);
   }
 
-  // Optimize the inference graph, and if needed, build the gradient graph.
   std::unordered_set<std::string> x_node_arg_names;
-  ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
-  if (!config_.build_gradient_graph) {
-    // This graph will be used only for inferencing, so stop right here.
-    // No need to build a gradient graph.
-    return Status::OK();
-  }
+  bool modified_graph = false;
+
+  do {
+    // Optimize the inference graph, and if needed, build the gradient graph.
+    ORT_RETURN_IF_ERROR(OptimizeInferenceGraph(x_node_arg_names));
+    if (!config_.build_gradient_graph) {
+      // This graph will be used only for inferencing, so stop right here.
+      // No need to build a gradient graph.
+      return Status::OK();
+    }
+
+    // expand any nodes that have an ONNX function definition but no matching ORT kernel.
+    modified_graph = false;
+    Graph& graph = gradient_model_->MainGraph();
+    ORT_RETURN_IF_ERROR(InlineNodes(graph, modified_graph));
```

Comment:
is there any case that we define the gradient builder for a function? Instead of blindly inline everything, it is better we check the gradient builder registry for each node first, and only inline the nodes that doesn't have gradient builder registered.

Reply:
> is there any case that we define the gradient builder for a function?
Yes, it could happen in case of custom ops where an EP claims a function and may have a gradient builder registered for it.
Instead of inlining all functions, I revised my PR to check if a) no EP claims the node and b) no gradient builder is registered for the node, before inlining the function node.